sklearn/metrics/tests/test_pairwise_distances_reduction.py

```diff
@@ -96,24 +182,524 @@ def assert_radius_neighborhood_results_equality(ref_dist, dist, ref_indices, ind
+        dist_row = dist[query_idx]
+
+        assert is_sorted(
+            ref_dist_row
+        ), f"Reference distances aren't sorted on row {query_idx}"
+        assert is_sorted(dist_row), f"Distances aren't sorted on row {query_idx}"
+
+        # Vectors' lengths might be different due to small
+        # numerical differences of distance w.r.t the `radius` threshold.
+        largest_row = ref_dist_row if len(ref_dist_row) > len(dist_row) else dist_row
+
+        # For the longest distances vector, we check that last extra elements
+        # that aren't present in the other vector are all in: [radius ± atol]
+        atol = 10 ** (-decimals)
+        min_length = min(len(ref_dist_row), len(dist_row))
+        last_extra_elements = largest_row[min_length:]
+        if last_extra_elements.size > 0:
+            assert np.all(radius - atol <= last_extra_elements <= radius + atol), (
+                f"The last extra elements ({last_extra_elements}) aren't in [radius ±"
+                f" atol]=[{radius} ± {atol}]"
+            )
+
+        # We truncate the neighbors results list on the smallest length to
+        # be able to compare them, ignoring the elements checked above.
+        ref_dist_row = ref_dist_row[:min_length]
+        dist_row = dist_row[:min_length]
+        print(type(ref_dist_row))
+        print(type(dist_row))
+
+        assert_allclose(ref_dist_row, dist_row, rtol=rtol)
+
+        ref_indices_row = ref_indices[query_idx]
+        indices_row = indices[query_idx]
+
+        # Grouping indices by distances using sets on
+        # a rounded distances up to a given number of decimals
+        reference_neighbors_groups = defaultdict(set)
+        effective_neighbors_groups = defaultdict(set)
+
+        for neighbor_rank in range(min_length):
+            rounded_dist = adaptive_rounding(
+                ref_dist_row[neighbor_rank], n_significant_digits=decimals
+            )
+            reference_neighbors_groups[rounded_dist].add(ref_indices_row[neighbor_rank])
+            effective_neighbors_groups[rounded_dist].add(indices_row[neighbor_rank])
+
+        # Asserting equality of groups (sets) for each distance
+        msg = (
+            f"Neighbors indices for query {query_idx} are not matching "
+            f"when rounding distances at decimals={decimals}"
+        )
+        for rounded_distance in reference_neighbors_groups.keys():
+            assert (
+                reference_neighbors_groups[rounded_distance]
+                == effective_neighbors_groups[rounded_distance]
+            ), msg
+
+
 ASSERT_RESULT = {
-    PairwiseDistancesArgKmin: assert_argkmin_results_equality,
-    PairwiseDistancesRadiusNeighborhood: assert_radius_neighborhood_results_equality,
+    # In the case of 64bit, we test for exact equality of the results rankings
+    # and standard tolerance levels for the computed distance values.
+    (PairwiseDistancesArgKmin, np.float64): assert_argkmin_results_equality,
```

Comment:
Note that in the future we might be interested in using quasi equality checks also for float64 data (with a larger number of significant digits) as the tests could be unstable because of numerically tied distances on some datasets (e.g. uniform grids).

So far we always wrote the neighbors tests to avoid this situation but this new help would allow us to write robust tests with expected numerical or exact ties.

Reply:
Nothing to do as part of this PR, this is just a remark.